blog/2022-05-13-caching-api-requests-with-long-lived-workflows.md

```diff
@@ -0,0 +1,231 @@
+  const ratesByDay = new Map<string, any>();
+
+  // Allow querying exchange rates by day
+  setHandler(getExchangeRatesQuery, (date: string) => {
+    return ratesByDay.get(date);
+  });
+
+  while (true) {
+    const exchangeRates = await getExchangeRates();
+    const today = new Date();
+    // Store today's exchange rates
+    ratesByDay.set(toYYYYMMDD(today), exchangeRates);
+    console.log(toYYYYMMDD(today), exchangeRates);
+
+    // Delete the oldest key if we have more than 30 entries
+    const keys = Array.from(ratesByDay.keys());
+    if (keys.length > maxNumRates) {
+      ratesByDay.delete(keys[0]);
+    }
+
+    // Wait until tomorrow at 12pm to refresh the exchange rates
+    const tomorrow = new Date(today);
+    tomorrow.setHours(12, 0, 0, 0);
+    tomorrow.setDate(tomorrow.getDate() + 1);
+    // @ts-ignore
+    await sleep(tomorrow - today);
+  }
+}
+```
+
+Temporal makes `ratesByDay` durable, even though `ratesByDay` is just a normal JavaScript variable.
+That's because Temporal stores the entire history of events for this Workflow.
+If the machine running `exchangeRatesWorkflow()` crashes, Temporal can resume the Workflow on another machine by replaying the entire event history.
+
+## Continue As New
+
+The `exchangeRatesWorkflow` can run for an unlimited period of time: days, months, even years.
+However, Temporal caps a Workflow at [50,000 events](https://docs.temporal.io/docs/temporal-explained/workflows/#:~:text=There%20is%20a%20hard%20limit,Workflow%20Execution%20is%20forcefully%20terminated.).
+In the `exchangeRatesWorkflow`, there are 4 events fired at every iteration of the `while` loop, assuming no API errors.
+
+1. `EVENT_TYPE_TIMER_FIRED`: the `setTimeout()` resolved and it's time to refresh the exchange rates
+2. `EVENT_TYPE_ACTIVITY_TASK_STARTED`: the `getExchangeRates()` activity started executing
+3. `EVENT_TYPE_ACTIVITY_TASK_COMPLETED`: the `getExchangeRates()` activity completed successfully
+4. `EVENT_TYPE_TIMER_STARTED`: the Workflow used `setTimeout()` to pause until tomorrow
+
+With 1 API request per day, the `exchangeRatesWorkflow()` can run for almost 12,500 days (approximately 34 years) before running into the 50,000 event limit.
+However, you should still handle this limit.
+And that's what Continue As New is for.
+
+You can think of Continue As New as restarting your Workflow with from an initial state.
+The only data that `exchangeRatesWorkflow()` needs to respond to queries is the `ratesByDay` map, so `exchangeRatesWorkflow()` needs to Continue As New with a serialized version of the `ratesByDay` map.
+The `exchangeRatesWorkflow()` also needs to be able to resume from a previous state.
+Continue As New just calls `exchangeRatesWorkflow()` with an initial state.
+Below is 
+
+```ts
+const maxNumRates = 30;
+const maxIterations = 10000;
+
+// `storedRatesByDay` contains the serialized data from Continue As New, if available. Otherwise, just an
```

Comment:
Ideally, change `Continue As New` to `Continue-As-New` here and in your repo (here and in the last comment in this excerpt).